protocols/manage/propagate.go

```diff
@@ -51,22 +47,48 @@ type PropagateReply struct {
 	Level int
 }
 
-// PropagateStartAndWait starts the propagation protocol and blocks until
+// PropagationFunc starts the propagation protocol and blocks until
 // all children stored the new value or the timeout has been reached.
 // The return value is the number of nodes that acknowledged having
 // stored the new value or an error if the protocol couldn't start.
-func PropagateStartAndWait(c *sda.Context, el *sda.Roster, msg network.Body, msec int, f func(network.Body)) (int, error) {
-	tree := el.GenerateNaryTreeWithRoot(8, c.ServerIdentity())
-	log.Lvl3("Starting to propagate", reflect.TypeOf(msg))
-	pi, err := c.CreateProtocolService("Propagate", tree)
-	if err != nil {
-		return -1, err
-	}
-	return propagateStartAndWait(pi, msg, msec, f)
+type PropagationFunc func(el *sda.Roster, msg network.Body, msec int) (int, error)
+
+// PropagationStore is the function that will store the new data.
+type PropagationStore func(network.Body)
+
+// NewPropagationFunc registers a new protocol name with the context c and will
+// set f as handler for every new instance of that protocol.
+func NewPropagationFunc(c *sda.Context, name string, f PropagationStore) (PropagationFunc, error) {
+	pid, err := c.ProtocolRegister(name, func(n *sda.TreeNodeInstance) (sda.ProtocolInstance, error) {
+		p := &Propagate{
+			sd:               &PropagateSendData{[]byte{}, 1000},
+			TreeNodeInstance: n,
+			received:         0,
+			subtree:          n.TreeNode().SubtreeCount(),
```

Comment:
Is this needed ? Does it go over all over the tree in order to compute that value ?